### ISSUE TITLE
Count is required when enumerable is not a collection

### ISSUE THREAD

**Issue by newsoftinc** [pieceofsummer/Hangfire.Console#10]
Extension method WithProgress will throw exception is collection has no items

else if (count < 0)
Count is required when enumerable is not a collection .

You receive and IEnumerable and return an IEnumerable implementation named progress enumerable.

Instead of throwing, could you just return the enumerable parameter ?

Suggested implementation :

```
public static IEnumerable WithProgress(this IEnumerable enumerable, IProgressBar progressBar, int count = -1)
        {
            if (enumerable is ICollection)
            {
                count = ((ICollection)enumerable).Count;
            }
            else if (count < 0)
            {
               return enumerable;
            }

            return new ProgressEnumerable(enumerable, progressBar, count);
        }

```

**Comment by pieceofsummer** [comment]
The purpose of that exception was supposed to guard against misusing progress bars with enumerables of unknown size. If we just return source enumerable, the progress bar will never be updated (will stay at zero), so it basically loses its purpose.

**Comment by newsoftinc** [comment]
That's what I was thinking, but it's kind of bad to break the execution stack to display a progression .. Maybe you can return a new instance of whatever progress enumerable and set its progress to 100%

**Comment by pieceofsummer** [comment]
It's quite rare to enumerate non-`ICollection` collections (I can only think of `yield` enumerables and custom enumerables), so requirement to pass `count` for those is rather ensuring that *you're aware* of enumerating a collection of unknown size.

`count` shouldn't necessary be the exact number of items, it's more like an estimation. If you can't (by any means) determine the source collection size, you can pass zero for `count` argument. It will still attempt to display an adequate progress then, even though its growth won't be linear.

**Comment by newsoftinc** [comment]
It's not rare to use typed arrays, which are implementations of IEnumerable that are not ICollection.  Just keep it as it is. I will use progress.SetValue. As mentionned earlier, my business process running is more important than an updated ui element breaking the stack.

**Comment by pieceofsummer** [comment]
Arrays implement both `ICollection` and `ICollection<T>`, so it shouldn't be an issue.

https://dotnetfiddle.net/Widget/K7bhu1

**Comment by newsoftinc** [comment]
That's correct. I was testing on an array filtered by a Where clause and did not know Where implementation on array returns a WhereArrayIterator instance. 
https://referencesource.microsoft.com/#system.core/System/Linq/Enumerable.cs,4158286b17727025

> 		Type[] arrType = new Type[] { typeof(System.Object) };
> 		var whereIterator = from V in arrType where V != null select V;
> 		var tryCast = whereIterator as ICollection<Type>;

**Comment by pieceofsummer** [comment]
I agree, there's a problem with LINQ operators, as they're lazy evaluated (basically, they're `yield` enumerables), so it is impossible to determine their size without actually enumerating them to the end.

You have two options here:

1. Determine count by enumerating it once, before actual processing:
```c#
var whereIterator = ...;
foreach (var item in whereIterator.WithProgress(bar, whereIterator.Count())) 
{
   // ...
}
```

2. Materialize collection before iterating it:
```c#
var whereIterator = ...;
foreach (var item in whereIterator.ToList().WithProgress(bar))
{
   // ...
}
```

First way is more memory-efficient, while the second can be faster when using complex LINQ operators.

**Comment by pieceofsummer** [comment]
I was considering to write 
```c#
    }
    else if (count < 0)
    {
        count = enumerable.Count();
    }
```
but I'm not sure that is always a better option rather than materialization, and even if any enumerator is guaranteed to be allowed enumerating twice, so it is better to specify that explicitly.

**Comment by ronaldbarendse** [comment]
It's probably better to use more specific `WithProgress` extension methods, e.g.:

```c#
public static IEnumerable<T> WithProgress<T>(this ICollection<T> collection, IProgressBar progressBar)
{
    return collection.WithProgress<T>(collection, progressBar, collection.Count);
}

public static IEnumerable<T> WithProgress<T>(this IEnumerable<T> enumerable, IProgressBar progressBar, int count)
{ }
```

So when using `ICollection`, it automatically uses the count, otherwise you just have to pass it manually (e.g. in case of an `IReadOnlyCollection`, adding this as additional method would make the signature ambiguous). And if `count` is less than 0, it should throw an `ArgumentOutOfRangeException`...
